third_party/move/move-compiler-v2/src/inliner.rs

```diff
@@ -0,0 +1,1184 @@
+        .iter()
+        .filter(|(_, callees)| !callees.is_empty())
+        .map(|(caller_fnid, _)| caller_fnid)
+        .cloned()
+        .collect();
+
+    // Check for cycles
+    let cycles = check_for_cycles(&inline_functions_with_callees);
+    if !cycles.is_empty() {
+        for cycle in cycles {
+            let start_fnid = cycle.first().unwrap();
+            let func_env = env.get_function(*start_fnid);
+            let path_string: String = cycle
+                .iter()
+                .map(|fnid| env.get_function(*fnid).get_name_str())
+                .collect::<Vec<String>>()
+                .join(" -> ");
+            let mut call_details: Vec<_> = cycle
+                .iter()
+                .zip(cycle.iter().skip(1).chain([*start_fnid].iter()))
+                .flat_map(|(f, g)| {
+                    let sites_ids = call_site_locations.get(&(*f, *g)).unwrap();
+                    let f_str = env.get_function(*f).get_name_str();
+                    let g_str = env.get_function(*g).get_name_str();
+                    let msg = format!("call from {} to {}", f_str, g_str);
+                    sites_ids
+                        .iter()
+                        .map(move |node_id| (env.get_node_loc(*node_id), msg.clone()))
+                })
+                .collect();
+            let msg = format!(
+                "recursion during function inlining not allowed: {} -> {}",
+                path_string,
+                func_env.get_name_str()
+            );
+            let loc = call_details.first_mut().unwrap().0.clone();
+            env.diag_with_labels(Severity::Error, &loc, &msg, call_details);
+        }
+        return Err(());
+    }
+
+    // Compute post-order of inlined_functions which call others.
+    let po_inline_functions = postorder(
+        &inline_functions_calling_others,
+        &inline_functions_with_callees,
+    );
+
+    // Identify subset of non-inline functions which call inline functions.  Order doesn't matter here.
+    let non_inline_functions_needing_inlining: Vec<QualifiedFunId> = function_callees
+        .iter()
+        .filter(|(fnid, callees)| {
+            !env.get_function(**fnid).is_inline()
+                && callees
+                    .iter()
+                    .any(|fnid2| env.get_function(*fnid2).is_inline())
+        })
+        .map(|(fnid, _callees)| fnid)
+        .cloned()
+        .collect();
+
+    let result: Vec<QualifiedFunId> =
+        chain(po_inline_functions, non_inline_functions_needing_inlining).collect();
+    Ok(result)
+}
+
+// Calculate a bottom-up traversal for entries, given the provided callee map.
+fn postorder<T: Ord + Copy + Debug>(
+    entries: &Vec<T>,
+    callee_map: &BTreeMap<T, BTreeSet<T>>,
+) -> Vec<T> {
+    let mut stack = Vec::new();
+    let mut visited = BTreeSet::new();
+    let mut grey = BTreeSet::new();
+    let mut postorder_num_to_node = Vec::new();
+    let mut node_to_postorder_num = BTreeMap::new();
```

Comment:
Remove this the operation on it, they don't contribute to anything.

Reply:
Ok.